### DISCUSSION TITLE
Invalidate all queries except one

### DISCUSSION THREAD

**Discussion by ryanmk54** [TanStack/query#2203] [Q&A] [answered]
I am writing an invoice search screen that allows modifying an invoice with a mutation. The mutation returns the updated invoice. This means I can update the currently search results without re-fetching the query. The same invoice could be in other search queries. Instead of checking all queries for the modified invoice, I was planning on invalidating all of the invoice queries, except the current one. If I use a similar key for the invoice queries, I can invalidate all of them with one invalidateQueries call. Is there a way to invalidate all of them except one?

The function used to update the current search query is below:
```
  const onUpdate = (invoice: ApInvoiceResult) => {
    // TODO all invoice queries except this one need to be invalidated
    queryClient.setQueryData(props.queryKey, (prev: ApInvoiceResult[]) =>
      prev.map(prevApInvoice => prevApInvoice.id === invoice.id ?
        deepmerge(prevApInvoice, invoice) : prevApInvoice
      )
    );
  };
```

**Comment by TkDodo** [comment, marked as answer]
there is. `invalidateQueries` also takes a predicate function as filter argument instead of the partial key. the example from the [docs](https://react-query.tanstack.com/guides/query-invalidation#query-matching-with-invalidatequeries) is:

```
 queryClient.invalidateQueries({
   predicate: query =>
     query.queryKey[0] === 'todos' && query.queryKey[1]?.version >= 10,
 })
```

if the function returns true, it will be invalidated, otherwise not.

---

you can achieve something similar with `setQueryData` if you don't want to invalidate by doing:

```
queryClient
  .getQueryCache()
  .findAll(filter)
  .forEach(({ queryKey }) => {
    queryClient.setQueryData(queryKey, newData)
  })
```

`findAll` takes the same filter (partial query key, predicate function) as invalidateQueries

**Reply by ryanmk54** [reply]
Awesome! Thank you. 😃
